### PR TITLE
build GSI utilities from GSI-utils repo

### PR DESCRIPTION
**Pull request by aerorahul**
**Description**
The GSI Utilities have been moved to https://github.com/NOAA-EMC/GSI-utils

This PR:
- checks out GSI utilities from the above repo
- builds the utilities from the above repo
- does not build the utilities from the GSI repo.  In a subsequent PR when the GSI removes the utilities, that option will become redundant.
- updates gsi-monitor hash and uses `build.sh` from that repository
- creates links appropriately.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [X] Clone and Build tests on Hera
- [ ] Cycled test on Orion
- [N/A] Forecast-only test on Hera
  
**Checklist**

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

File: sorc/link_workflow.sh
Comment (RussTreadon-NOAA) [inline sorc/link_workflow.sh:292]: **Two comments.**
1) We are still in `exec/` when this section is executed.  We should move line 298, namely, 

    `cd ${pwd}/../sorc   ||   exit 8`

    to just before line 289.   Otherwise the ndiag, global_gsi and global_enkf source directories are created in the wrong location.

2)  I think operational Implementation standards require the source code and executable name to be consistent.  For example, if the executable is `gsi.x`, the source code directory should be `gsi.fd`.   If this is true, we need to change the source code directory names in this section to be consistent with the executable name.

Comment (aerorahul) [reply sorc/link_workflow.sh:292]: Thanks @RussTreadon-NOAA for catching the linking of the directories.
I have retained the names of the executables to be consistent with the source code.  `ncdiag.fd` will become moot soon.

Comment (RussTreadon-NOAA) [reply sorc/link_workflow.sh:292]: It dawned on me that we already have `gsi.fd` from `checkout.sh`.   Thus, the following section of `link_workflow.sh` is not needed
```
    [[ -d gsi.fd ]] && rm -rf gsi.fd
    $SLINK gsi.fd/src/gsi     gsi.fd
```
In fact, retaining the above erroneously removes the clone of NOAA-EMC/GSI in `gsi.fd`.

**Review by RussTreadon-NOAA — COMMENTED** [review]
Changes look good apart from removal of `gsi.fd` link from `link_workflow.sh`.   My previous comment prompted this change.  I was wrong and corrected myself in a subsequent comment.  Assuming `link_workflow.sh` is fixed, this set of changes is ready for merger into g-w develop.  

Once GSI PR [#427](https://github.com/NOAA-EMC/GSI/pull/427) is merged into GSI develop this g-w PR or a subsequent g-w PR needs to  
1) clean up the _TODO_ section of `build_gsi.sh`
2) update the hash for GSI develop in `checkout.sh` to the GSI PR [#427](https://github.com/NOAA-EMC/GSI/pull/427) hash or one thereafter.

File: .gitignore
Comment (RussTreadon-NOAA) [inline .gitignore:85]: Lines 105 and 106 reference `global_enkf.fd` and `global_gsi.fd`.   These directories are not created by the updated `link_workflow.sh`.   Instead we have `gsi.fd` and `enkf.fd`.     We can remove `global_gsi.fd` and `global_enkf.fd`.   `gsi.fd` is already in `.gitignore`.  We can add `enkf.fd`.   

For the time being `link_workflow.sh` creates `ncdiag.fd`.   For completeness we could replace `ncdiag_cat.fd` in `.gitignore` with `ncdiag.fd`

Comment (aerorahul) [reply .gitignore:85]: That raises an interesting question.
`gsi.fd` from `checkout.sh` is the "umbrella" GSI containing the source codes for `GSI`, `EnKF` and soon to be removed `ncdiag`.
Do (Should) we need to "link" `GSI/src/gsi` and `GSI/src/enkf` to `sorc`?
EE2 standards say yes.
What if we call the `gsi.fd` from `checkout.sh` as `GSIApp.fd` and then retain the symlinks?
Just a thought for now and can be worked later.
I will pin this for discussion w/ NCO.

Comment (aerorahul) [reply .gitignore:85]: @RussTreadon-NOAA 
See if you agree with [5457fe0](https://github.com/NOAA-EMC/global-workflow/pull/889/commits/5457fe052524fbe65c87f90ccdf5aea54a70b36d) for the time being.
`gsi_enkf.fd` is accurate to what contains in that directory.

Comment (RussTreadon-NOAA) [reply .gitignore:85]: Cloning NOAA-EMC/GSI into `GSIApp.fd` and updating links would work.  If we take this approach we may want to similarly clone NOAA-EMC/GDASApp into `GDASApp.fd` and update JEDI links accordingly.   

I agree.  We don't need to answer all these questions in this particular PR.   We simply need this PR to build GSI utilities from the GSI-utils repo and update links accordingly.

File: sorc/build_gsi.sh
Comment (RussTreadon-NOAA) [inline sorc/build_gsi.sh]: Given changes to `checkout.sh` need to replace `gsi.fd` with `gsi_enkf.fd`.  For clarity we should rename `build_gsi.sh` as `build_gsi_enkf.sh`

File: sorc/build_all.sh
Comment (RussTreadon-NOAA) [inline sorc/build_all.sh:136]: build_gsi section above is now more accurately described as build_gsi_enkf. If we rename `build_gsi.sh` as `build_gsi_enkf.sh`, corresponding changes are needed above.

**Comment by RussTreadon-NOAA** [comment]
After updating to [0cc159d](https://github.com/NOAA-EMC/global-workflow/pull/889/commits/0cc159d3cb2bf4a0aacd73588a59d810b74c933e) found it necessary to update  `gfs_build.cfg` as follows
```
- Building gsi (gsi) .................................... yes
+ Building gsi_enkf (gsi_enkf) .......................... yes
```
Absent this change `build_all.sh` aborts with
```
Orion-login-1[48] rtreadon$ ./build_all.sh
*** Unrecognized line in config file "gfs_build.cfg":
Building gsi (gsi) .................................... yes
```

**Comment by aerorahul** [comment]
> After updating to [0cc159d](https://github.com/NOAA-EMC/global-workflow/pull/889/commits/0cc159d3cb2bf4a0aacd73588a59d810b74c933e) found it necessary to update `gfs_build.cfg` as follows
> 
> ```
> - Building gsi (gsi) .................................... yes
> + Building gsi_enkf (gsi_enkf) .......................... yes
> ```
> 
> Absent this change `build_all.sh` aborts with
> 
> ```
> Orion-login-1[48] rtreadon$ ./build_all.sh
> *** Unrecognized line in config file "gfs_build.cfg":
> Building gsi (gsi) .................................... yes
> ```

Ok.
I see whats happening here.
This is incredibly complicated set of logic and jumping through hoops for something that shouldn't be this convoluted.

**Comment by RussTreadon-NOAA** [comment]
Sorry, a bit more hoop jumping.

Execution of `./build_all.sh` revealed a problem with gsi_monitor.

```
Orion-login-1[55] rtreadon$ ./build_all.sh
 .... Building WW3 pre and post execs ....
 .... Building forecast model ....
 .... Building gsi and enkf ....
 .... Building gsi utilities ....
 .... Building gsi monitor ....
Fatal error in building gsi monitor.
The log file is in /work/noaa/da/Russ.Treadon/git/global_workflow/pr889/sorc/logs/build_gsi_monitor.log
```

We need to replace `gsi.fd` with `gsi_enkf.fd` in `build_gsi_monitor.sh`

```
-GSI_INSTALL_PREFIX="${cwd}/gsi.fd/install"
+GSI_INSTALL_PREFIX="${cwd}/gsi_enkf.fd/install"
```

Also updated error message in `build_gsi_monitor.sh`
 ```
else
   echo << EOF
 FATAL: ${GSI_INSTALL_PREFIX} does not exist
-       Have you built GSI yet?
+       Have you built GSI and EnKF yet?
 EOF
   exit 2
```

**Comment by aerorahul** [comment]
@RussTreadon-NOAA 
Let me make sure this all works before you review again.
I don't want you to troubleshoot errors and mistakes.
I can close the PR, if that helps.

**Comment by RussTreadon-NOAA** [comment]
@aerorahul , I think we're at the end of the build issues for this PR.   `build_all.sh` just finished in my working directory.   It ran to completion.    

The only potential issue for `link_workfow.sh` is whether or not we want to add `sorc/enkf.fd`, `sorc/gsi.fd`, and `sorc/ncdiag.fd` to `.gitignore`.   _git status_ returns these three directories as _Untracked files_.    `sorc/ncdiag.fd` will be removed in the future.   `enkf.fd` and `gsi.fd` will remian.

**Comment by aerorahul** [comment]
I have now, done all the steps:
`checkout.sh`,
`build_all.sh`, 
`link_workflow.sh`
and I don't see errors or stale links.

File: sorc/build_gsi_enkf.sh
Comment (WalterKolczynski-NOAA) [inline sorc/build_gsi_enkf.sh]: Default should be NO so the -v option actually does something, but I think having it at all is unnecessary since the default if not set is no anyway.

File: sorc/build_all.sh
Comment (WalterKolczynski-NOAA) [inline sorc/build_all.sh:15]: Still don't approve of this change that breaks the nice formatting.

File: sorc/build_gsi_monitor.sh
Comment (WalterKolczynski-NOAA) [inline sorc/build_gsi_monitor.sh:4]: I think this may need to stay to support partial build? If not, we should take it out of all the build scripts.

**Review by RussTreadon-NOAA — APPROVED** [review]
Clone `feature/gsi-utils` at 4db8b89d.  Run `checkout.sh`, `build_all.sh`, and `link_workflow.sh`.   All three scripts ran without error.  A quick check of g-w directories does not show any incomplete links.  _git status_ does not return any untracked files.

**Comment by aerorahul** [comment]
Waiting on @WalterKolczynski-NOAA to review.

File: sorc/build_gsi_monitor.sh
Comment (WalterKolczynski-NOAA) [inline sorc/build_gsi_monitor.sh]: Default should be "NO"

File: sorc/build_gsi_monitor.sh
Comment (WalterKolczynski-NOAA) [inline sorc/build_gsi_monitor.sh:37]: Does this line do anything?

File: sorc/build_gsi_utils.sh
Comment (WalterKolczynski-NOAA) [inline sorc/build_gsi_utils.sh]: As before

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: sorc/build_gsi_monitor.sh
Comment (aerorahul) [reply sorc/build_gsi_monitor.sh:4]: Can you verify it is needed for partial build?
I tried with a partial build for the GSI, GSI-utils and GSI-monitor and had no impact.
I think it is necessary for builds that loads modules in the build scripts in this repo.
It is not necessary for GSI, GSI-utils or GSI-monitor that uses the `build.sh` from their repos.

File: sorc/build_gsi_monitor.sh
Comment (aerorahul) [reply sorc/build_gsi_monitor.sh:37]: Yes.  It passes `-Dncdiag_ROOT=${GSI_INSTALL_PREFIX}` to the gsi-monitor cmake command.

File: sorc/build_all.sh
Comment (aerorahul) [reply sorc/build_all.sh:15]: formatting is preserved.

File: sorc/build_gsi_monitor.sh
Comment (WalterKolczynski-NOAA) [reply sorc/build_gsi_monitor.sh:4]: Got it

**Review by WalterKolczynski-NOAA — COMMENTED** [review]
I still protest the conversion to spaces that breaks the heredoc indentation, but will let it through if I must. I want to wait until after the cycling PR gets merged, since that is imminent, more important, and modifies the same area. This one will have to be slightly modified to bring in those changes.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Pull request merged**
